### ISSUE TITLE
Tables (string, data, search) takes magnitudes longer to load/filter for newer smaller binary compared to older larger binary from same program

### ISSUE THREAD

**Issue by I-Knight-I** [NationalSecurityAgency/ghidra#3460]
**Describe the bug**
The Defined Strings, Data and String Search tables take magnitudes longer to load for a newer smaller binary compared to an older larger binary of the same program. 

The former will take approx. 10 - 15 minutes to load ~50,000 - 100,000 items 
The latter will take approx. 5 - 10 seconds to load ~50,000 - 120,000 items

Filtering items takes approx. the same amount of time. 

**To Reproduce**
*Currently unknown*
Need help/suggestions for investigating

**Expected behavior**
Ghidra should populate the tables of the smaller binary quicker than or at the same speed of its older and larger counter-part. 

**Screenshots**
Binary breakdown:
![image](https://user-images.githubusercontent.com/26165117/135073745-5adce690-f622-4119-99dc-41a8b5715e61.png)

**Attachments**
N/A

**Environment (please complete the following information):**
 - OS: Windows 10
 - Java Version: Java 16.0.1 [build: 16.0.1+9-24]
 - Ghidra Version: 10.0.3 (but has persisted since at least 9.2)
 - Ghidra Origin: Repo's releases tab

**Additional context**
I'm unable to legally share the executable files but am happy to provide snippets of information and logs where possible.

**Comment by dragonmacher** [comment]
I love a good performance problem.   In lieu of the binary, periodic stack traces may help us diagnose the issue.   You can run these commands to generate thread trace information:
`jps`
`jstack <pid`

The `jps` command will give you a list of Java pids.  Feed the Ghidra pid into the `jstack` command.   A few stack dumps while the tables are loading will be helpful.

This issue is peculiar, as quite a bit of effort has gone into performance improvements in this area.

**Comment by I-Knight-I** [comment]
[stack_dumps.zip](https://github.com/NationalSecurityAgency/ghidra/files/7244927/stack_dumps.zip)

I took a good few from the new binary. Hopefully they'll be of use. 
Lemme know if you'd like them as a gist or uploaded as individual .txt instead of the .zip.

Not sure if unrelated or not but the defined strings table fully loaded at 47%. 
Defined Data also jumped to 10% when it started for both populating and sorting. 

You're right though, I've been keeping tabs on other PRs regarding defined strings performance. I sat on this issue since I first used Ghidra on version 9.2 but was hoping it'd be fixed.

**Comment by I-Knight-I** [comment]
That's definitely not right, I'm not touching it at all 😄 
No scripts are running and, whilst I was recording those dumps for you, I wasn't even navigating through the listing, functions, etc. I didn't touch Ghidra at all.

**Comment by dragonmacher** [comment]
OK.  I'll trying to dig up a comparable binary to see if I can reproduce the issue.

**Comment by dragonmacher** [comment]
Also, did you run both versions of Ghidra with the same version of Java?

**Comment by I-Knight-I** [comment]
Yup. Both should've used Java 16 by default since that's the runtime on my path.

**Comment by dragonmacher** [comment]
For the tables in question, can you list for me the name of each column in the table?

**Comment by I-Knight-I** [comment]
Defined Strings:
- Location
- String Value
- String Representation
- Data Type
Defined Data:
- Data
- Location
- Type
- Size
String Search:
- Defined
- Location
- Label
- Code Unit
- String View
- String Type
- Length
- Is Word

**Comment by dragonmacher** [comment]
I just tested a binary with 640k define data items.  For the Defined Data table, the loading took about 1 minute, sorting took about 1 minute and filtering took about 1 minute.

The Defined Strings table had roughly the same numbers.   The String Search seemed about the same, maybe a bit quicker.

I did not noticed any repeated sorting, as is implied by your trace files.    If that is happening, that could explain the times your are seeing.

**Comment by I-Knight-I** [comment]
That sounds likely. 
How can I verify and/or forcefully stop the table from repeatedly sorting? 

Again, I can assure you that no analysis or scripts are running and I'm not making any edits to the binary itself. So I'm not sure where to go from here

**Comment by dragonmacher** [comment]
Well, as an experiment, you *should* be able to `Control-click` on the current sort column in the Defined Data table to remove the sort.   (You can stop any loading/sorting before you do this.)   After the sort has been removed, close  Ghidra and restart.   Assuming you tool reopens with no column showing as sorted, then this should prevent any sorting from taking place.

**Comment by I-Knight-I** [comment]
It seems hit or miss with `Control-click`. 
I think it loaded once without sorting enabled and it was still the same speed. Every other time, a column had the sorting symbol next to it.

**Comment by dragonmacher** [comment]
I'm pouring over the traces and do not see anything out of place, other than the repeated sorting.

Have you checked the application log file to look for exception stack traces?   That file lives at {user home}/.ghidra/.ghidra_{version}/application.log.   It looks like the table loading state machine logs errors and then tries to recover.   There may be something odd happening there.

**Comment by I-Knight-I** [comment]
No exceptions. 

Began loading strings and defined data a few times just to be sure too and still nothing.

**Comment by dragonmacher** [comment]
OK.  I am out of ideas at this point.   If we can reproduce the issue, then we should be able to fix it.    If you find a shareable binary that exhibits the same behavior, then let us know.

**Comment by I-Knight-I** [comment]
Will do. Cheers for your time regardless.

**Comment by dev747368** [comment]
Also, what does the Help (in the project front end window)  | Show VM Memory show?

**Comment by dev747368** [comment]
Thats good enough.  Its showing that your high-water mark for memory usage (650Mb) didn't even approach the max available to the jvm (4g)

**Comment by dev747368** [comment]
(but maybe a shot of memory with the older / bigger / faster binary would be good also)

**Comment by I-Knight-I** [comment]
![image](https://user-images.githubusercontent.com/26165117/135140661-37cb866c-ef69-4b22-98df-adb3d3d76d07.png)

This time only with the older/bigger/faster binary tool opened. 
That's a screenshot within 30 seconds of re-opening Ghidra - I don't know if that'll skew the results at all since the slower binary's memory photo was (obviously) taken after all strings had loaded ~15-20+ minutes after opening. 

![image](https://user-images.githubusercontent.com/26165117/135140917-d19457a3-f1ae-480b-ae5d-edef5f21b37f.png)
That's the final value after leaving it idle as I wrote this response.
